lib/perl/Genome/Utility/File/DirCompare.pm

```diff
@@ -0,0 +1,289 @@
+#This module was adapted from File::DirCompare
+#(http://search.cpan.org/~gavinc/File-DirCompare-0.7/DirCompare.pm)
+#If two paths are both symlinks, this module will resolve the symlinks and do
+#a file comparison on the targets. The previous behavior in File::DirCompare
+#for two symlinks was to count them as different if they were pointing to
+#different targets. With this change they will only be counted as different if
+#the contents of the targets are different.
+
+package Genome::Utility::File::DirCompare;
+
+use strict;
+use warnings;
+
+use Genome;
+use File::Basename;
+use File::Spec::Functions;
+use File::Compare ();
+use File::Glob qw(bsd_glob);
+
+# ----------------------------------------------------------------------------
+# Private methods
+
+sub _dir_compare {
+    my $self = shift;
+    my ($dir1, $dir2, $sub, $opts) = @_;
+
+    # Glob $dir1 and $dir2
+    my (%d1, %d2);
+    $d1{basename $_} = 1 foreach bsd_glob(catfile($dir1, ".*"));
+    $d1{basename $_} = 1 foreach bsd_glob(catfile($dir1, "*"));
+    $d2{basename $_} = 1 foreach bsd_glob(catfile($dir2, ".*"));
+    $d2{basename $_} = 1 foreach bsd_glob(catfile($dir2, "*"));
+
+    # Prune dot dirs
+    delete $d1{curdir()} if $d1{curdir()};
+    delete $d1{updir()}  if $d1{updir()};
+    delete $d2{curdir()} if $d2{curdir()};
+    delete $d2{updir()}  if $d2{updir()};
+
+    # Setup cmp and matches subs
+    my $cmp = $opts->{cmp} && ref $opts->{cmp} eq 'CODE' ? $opts->{cmp} : \&File::Compare::compare;
+    my $matches = $opts->{matches} if $opts->{matches} && ref $opts->{matches} eq 'CODE';
+
+    # Iterate over sorted and uniquified file list
+    my %u;
+    for my $f (map { $u{$_}++ == 0 ? $_ : () } sort(keys(%d1), keys(%d2))) {
+        my $f1 = catfile($dir1, $f);
+        my $f2 = catfile($dir2, $f);
+        # Only in $dir1
+        if (! $d2{$f}) {
+            $sub->($f1, undef) unless $opts->{ignore_unique};
+        }
+        # Only in $dir2
+        elsif (! $d1{$f}) {
+            $sub->(undef, $f2) unless $opts->{ignore_unique};
+        }
+        # Item exists in both directories
+        else {
+            # Both symlinks
+            if (-l $f1 && -l $f2) {
+                my $t1 = readlink $f1 or die $self->error_message("Cannot read symlink $f1: $!");
```

Comment:
I have a local fix for this which uses `File::Find` with `follow` enabled instead of readlink. I also added tests for "normal" symlinks, symlinks to symlinks, and circular symlinks. I didn't push the changes since I will close this PR and look into implementing this tool via the genome vendor route.